### PR TITLE
State ofed impl

### PR DESCRIPTION
**Pull request by adrianchiris**

### CODE REVIEW

File: manifests/stage-ofed-driver/0010_ofed-driver-ds.yaml [hunk 1]
```diff
@@ -41,6 +42,5 @@ spec:
           hostPath:
             path: /run/mellanox/drivers
       nodeSelector:
-        #node-role.kubernetes.io/worker: ""
-        feature.node.kubernetes.io/pci-15b3_1017.present: "true"
-        feature.node.kubernetes.io/kernel-version.full: "5.4.0-29-generic"
\ No newline at end of file
+        feature.node.kubernetes.io/pci-15b3.present: "true"
+        feature.node.kubernetes.io/kernel-version.full: {{ .RuntimeSpec.KernelVerFull }}
```
Comment (moshe010) [inline manifests/stage-ofed-driver/0010_ofed-driver-ds.yaml:46]: newline

File: pkg/nodeinfo/attributes.go [hunk 2]
```diff
@@ -19,7 +19,7 @@ const (
 	nodeLabelKernelVerFull = "feature.node.kubernetes.io/kernel-version.full"
 	nodeLabelHostname      = "kubernetes.io/hostname"
 	nodeLabelCPUArch       = "kubernetes.io/arch"
-	nodeLabelMlnxNIC       = "feature.node.kubernetes.io/pci-15b3.present"
+	NodeLabelMlnxNIC       = "feature.node.kubernetes.io/pci-15b3.present"
```
Comment (moshe010) [inline pkg/nodeinfo/attributes.go:22]: You need it out side of the package?

Comment (adrianchiris) [reply pkg/nodeinfo/attributes.go:22]: Yes, as nodeinfo.Provider gets created with a list of nodes, the using entitiy need to get node objects from k8s API.

Now, as we dont want to get all the nodes in the cluster, we use labels to filter the `List` request.

The rest is used internally.
This kinda looks crooked, i agree. would you prefer to make them all public or separate into public/internal labels ?

Comment (moshe010) [reply pkg/nodeinfo/attributes.go:22]: let put all public even if you don't used them

File: pkg/state/state_ofed.go [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+func NewStateOFED(k8sAPIClient client.Client, scheme *runtime.Scheme, manifestDir string) (State, error) {
+	files, err := utils.GetFilesWithSuffix(manifestDir, render.ManifestFileSuffix...)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get files from manifest dir")
+	}
+
+	renderer := render.NewRenderer(files)
+	return &stateOFED{
+		stateSkel: stateSkel{
+			name:        "state-OFED",
```
Comment (moshe010) [inline pkg/state/state_ofed.go:29]: move to const

File: pkg/state/state_ofed.go [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+	CrSpec      *mellanoxv1alpha1.OFEDDriverSpec
+	RuntimeSpec *runtimeSpec
+}
+
+// Sync attempt to get the system to match the desired state which State represent.
+// a sync operation must be relatively short and must not block the execution thread.
+func (s *stateOFED) Sync(customResource interface{}, serviceCatalog ServiceCatalog) (SyncState, error) {
+	cr := customResource.(*mellanoxv1alpha1.NicClusterPolicy)
+	log.V(consts.LogLevelInfo).Info(
+		"State: OFED-driver Sync Custom resource", "Name", cr.Name, "Namespace:", cr.Namespace)
```
Comment (moshe010) [inline pkg/state/state_ofed.go:58]: you can use the name of state const above

File: pkg/state/state_ofed.go [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+	// Fill ManifestRenderData and render objects
+	nodeInfo := serviceCatalog.GetNodeInfoProvider()
+	if nodeInfo == nil {
+		return SyncStateError, errors.New("unexpected state, catalog does not provide node information")
+	}
+	objs, err := s.getManifestObjects(cr, nodeInfo)
+	if err != nil {
+		return SyncStateNotReady, errors.Wrap(err, "failed to create k8s objects from manifest")
+	}
+	if len(objs) == 0 {
```
Comment (moshe010) [inline pkg/state/state_ofed.go:76]: when will this happen?

File: pkg/state/state_skel.go [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+func (s *stateSkel) Name() string {
+	return s.name
+}
+
+// Description provides the State description
+func (s *stateSkel) Description() string {
+	return s.description
+}
+
+func (s *stateSkel) createOrUpdateObjs(
```
Comment (moshe010) [inline pkg/state/state_skel.go:42]: can we have 
1. createObjs()
2. updateObjs()
3. createOrUpdateObsj - we should find a better name (maybe SetObjs)

File: pkg/state/state_skel.go [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+		} else if err != nil {
+			// error occurred fetching object
+			return errors.Wrapf(
+				err, "failed to get object: Name: %s, Namespace: %s", obj.GetName(), obj.GetNamespace())
+		}
+
+		log.V(consts.LogLevelInfo).Info("Object found, Updating object.")
+		// Object found, Update it
+		// Note: Some objects may require update of the resource version
+		// TODO: using Patch preserves runtime attributes. consider using patch if relevant
```
Comment (moshe010) [inline pkg/state/state_skel.go:72]: mmm , to for my comment we should have
1. create (only if not exist) POST
2. update (patch update partial object) PATCH
3. set(create or update the object)  PUT

File: pkg/state/state_skel.go [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+		required := obj.DeepCopy()
+
+		if err := s.client.Update(context.TODO(), required); err != nil {
```
Comment (moshe010) [inline pkg/state/state_skel.go:75]: what is the required  is it the update objects? should we print them in debug?

File: pkg/state/state_skel.go [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+
+// Iterate over objects and check for their readiness
+func (s *stateSkel) getSyncState(objs []*unstructured.Unstructured) (SyncState, error) {
+	log.V(consts.LogLevelInfo).Info("Checking related object states")
+	for _, obj := range objs {
+		log.V(consts.LogLevelInfo).Info("Checking object", "Kind:", obj.GetKind(), "Name", obj.GetName())
+		// Check if object exists
+		found := obj.DeepCopy()
+		err := s.client.Get(
+			context.TODO(), types.NamespacedName{Name: obj.GetName(), Namespace: obj.GetNamespace()}, found)
```
Comment (moshe010) [inline pkg/state/state_skel.go:91]: maybe mobe this to stateSkel.getObjs()

File: pkg/state/state_skel.go [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+			log.V(consts.LogLevelInfo).Info("Object is not ready", "Kind:", obj.GetKind(), "Name", obj.GetName())
+			if k8serrors.IsNotFound(err) {
+				// does not exist (yet)
+				return SyncStateNotReady, nil
+			}
+			// other error
+			return SyncStateNotReady, errors.Wrapf(err, "failed to get object")
+		}
+		// Object exists, check for Kind specific readiness
+		if found.GetKind() == "DaemonSet" {
```
Comment (moshe010) [inline pkg/state/state_skel.go:102]: do we have constant for DaemonSet?

File: pkg/state/state_skel.go [hunk 11]
```diff
@@ -0,0 +1,147 @@
...
+				return SyncStateNotReady, err
+			}
+		}
+		log.V(consts.LogLevelInfo).Info("Object is ready", "Kind:", obj.GetKind(), "Name", obj.GetName())
+	}
+	return SyncStateReady, nil
+}
+
+// daemonsetReady checks if daemonset is ready
+func (s *stateSkel) daemonsetReady(uds *unstructured.Unstructured) (bool, error) {
```
Comment (moshe010) [inline pkg/state/state_skel.go:114]: s/daemonsetReady/isDeamonSetReady

File: pkg/state/state_skel.go [hunk 12]
```diff
@@ -0,0 +1,147 @@
...
+
+	log.V(consts.LogLevelDebug).Info(
+		"Check daemonset state",
+		"DesiredNodes:", ds.Status.DesiredNumberScheduled,
+		"CurrentNodes:", ds.Status.CurrentNumberScheduled,
+		"PodsAvailable:", ds.Status.NumberAvailable,
+		"PodsUnavailable:", ds.Status.NumberUnavailable,
+		"PodsReady:", ds.Status.NumberReady,
+		"Conditions:", ds.Status.Conditions)
+	if ds.Status.DesiredNumberScheduled != 0 && ds.Status.DesiredNumberScheduled == ds.Status.NumberAvailable {
```
Comment (moshe010) [inline pkg/state/state_skel.go:133]: why do we need ds.Status.DesiredNumberScheduled != 0?
if ds.Status.DesiredNumberScheduled  ==0 then ds.Status.NumberAvailable and that should be also done right?
if ds.Status.DesiredNumberScheduled  ==0  you will return false

File: pkg/state/state_ofed.go (hunk 5 above)
Comment (adrianchiris) [reply pkg/state/state_ofed.go:76]: it should error here

File: pkg/state/state_skel.go [hunk 13]
```diff
@@ -0,0 +1,147 @@
...
+func (s *stateSkel) createOrUpdateObjs(
+	setControllerReference func(obj *unstructured.Unstructured) error,
+	objs []*unstructured.Unstructured) error {
+	for _, obj := range objs {
+		log.V(consts.LogLevelInfo).Info("Handling manifest object", "Kind:", obj.GetKind(), "Name", obj.GetName())
+		// Set controller reference for object to allow cleanup on CR deletion
+		if err := setControllerReference(obj); err != nil {
+			return errors.Wrap(err, "failed to set controller reference for object")
+		}
+		// Check if this object already exists
```
Comment (adrianchiris) [inline pkg/state/state_skel.go:51]: add comment here to explain why we are deep copying in
add TODO to revisit it and perhaps just create an unstructured with just the object and type metadata

File: pkg/state/state_skel.go [hunk 14]
```diff
@@ -0,0 +1,147 @@
...
+
+// Description provides the State description
+func (s *stateSkel) Description() string {
+	return s.description
+}
+
+func (s *stateSkel) createOrUpdateObjs(
+	setControllerReference func(obj *unstructured.Unstructured) error,
+	objs []*unstructured.Unstructured) error {
+	for _, obj := range objs {
```
Comment (adrianchiris) [inline pkg/state/state_skel.go:45]: rename obj to desiredObj

File: pkg/state/state_skel.go (hunk 8 above)
Comment (adrianchiris) [reply pkg/state/state_skel.go:75]: no need for debug imo, we already print the rendered objects

File: pkg/state/state_skel.go (hunk 9 above)
Comment (adrianchiris) [reply pkg/state/state_skel.go:91]: yes i will separate to Get,Create,Update object methods and reuse throughout

File: pkg/state/state_skel.go (hunk 10 above)
Comment (adrianchiris) [reply pkg/state/state_skel.go:102]: no constant, i can define one but its only used here for the moment

File: pkg/state/state_skel.go (hunk 12 above)
Comment (adrianchiris) [reply pkg/state/state_skel.go:133]: if we get the object from k8s right after creating it, the initial status values are zero as the daemonset controller didnt update the fields yet.

there is an edge case where there are no suitable nodes for the daemonset to be scheduled.

I will check the behaviour

Comment (adrianchiris) [reply pkg/state/state_skel.go:133]: Added a note for now explaining whats going on and a TODO to check for alternatives.

File: pkg/state/state_skel.go (hunk 10 above)
Comment (adrianchiris) [reply pkg/state/state_skel.go:102]: I dont think a constant is needed here, also to my knowledge K8s libs dont provide it.

**Pull request merged**